src/Covenant/Internal/Expr.hs

```diff
@@ -0,0 +1,268 @@
+    prim,
+    app,
+    idOf,
+  )
+where
+
+import Control.Monad.Reader (Reader, ask, local, runReader)
+import Control.Monad.State.Strict (State, get, put)
+import Control.Monad.Trans (lift)
+import Covenant.Constant (AConstant)
+import Covenant.Prim (OneArgFunc, SixArgFunc, ThreeArgFunc, TwoArgFunc)
+import Data.Bimap (Bimap)
+import Data.Bimap qualified as Bimap
+import Data.Kind (Type)
+import Data.Word (Word64)
+import Test.QuickCheck (Arbitrary (arbitrary), chooseEnum)
+import Test.QuickCheck.GenT
+  ( GenT (GenT),
+    frequency,
+    liftGen,
+    oneof,
+    runGenT,
+    sized,
+  )
+
+-- | A unique identifier for a node in a Covenant program.
+--
+-- @since 1.0.0
+newtype Id = Id Word64
+  deriving
+    ( -- | @since 1.0.0
+      Eq,
+      -- | @since 1.0.0
+      Ord
+    )
+    via Word64
+  deriving stock
+    ( -- | @since 1.0.0
+      Show
+    )
+
+-- | An argument passed to a function in a Covenant program.
+--
+-- @since 1.0.0
+newtype Arg = Arg Word64
+  deriving
+    ( -- | @since 1.0.0
+      Eq,
+      -- | @since 1.0.0
+      Ord
+    )
+    via Word64
+  deriving stock
+    ( -- | @since 1.0.0
+      Show
+    )
+
+-- | A general reference in a Covenant program. This is either a computation
+-- (represented by its unique 'Id') or a function argument (represented by an
+-- 'Arg').
+--
+-- @since 1.0.0
+data Ref = AnArg Arg | AnId Id
+  deriving stock
+    ( -- | @since 1.0.0
+      Eq,
+      -- | @since 1.0.0
+      Ord,
+      -- | @since 1.0.0
+      Show
+    )
+
+-- | A call to a Plutus primitive.
+--
+-- @since 1.0.0
+data PrimCall
+  = PrimCallOne OneArgFunc Ref
+  | PrimCallTwo TwoArgFunc Ref Ref
+  | PrimCallThree ThreeArgFunc Ref Ref Ref
+  | PrimCallSix SixArgFunc Ref Ref Ref Ref Ref Ref
+  deriving stock
+    ( -- | @since 1.0.0
+      Eq,
+      -- | @since 1.0.0
+      Ord,
+      -- | @since 1.0.0
+      Show
+    )
+
+-- | A node in a Covenant program.
+--
+-- @since 1.0.0
+data Expr
+  = Lit AConstant
+  | Prim PrimCall
+  | Lam Ref
+  | App Ref Ref
+  deriving stock
+    ( -- | @since 1.0.0
+      Eq,
+      -- | @since 1.0.0
+      Ord,
+      -- | @since 1.0.0
+      Show
+    )
+
+-- | A helper monad for building up Covenant programs. In particular, this
+-- enables hash consing.
+--
+-- @since 1.0.0
+newtype ExprBuilder (a :: Type) = ExprBuilder (State (Bimap Id Expr) a)
+  deriving
+    ( -- | @since 1.0.0
+      Functor,
+      -- | @since 1.0.0
+      Applicative,
+      -- | @since 1.0.0
+      Monad
+    )
+    via (State (Bimap Id Expr))
+
+-- | Does not shrink.
+--
+-- @since 1.0.0
+instance Arbitrary (ExprBuilder Id) where
+  {-# INLINEABLE arbitrary #-}
+  arbitrary = do
+    generated <- runGenT (sized go)
+    pure $ runReader generated 0
+    where
+      go :: Int -> GenT (Reader Word64) (ExprBuilder Id)
+      go size
+        | size <= 0 = aLiteral
+        | otherwise = oneof [aLiteral, aPrim size, aLam size, anApp size]
+      aLiteral :: GenT (Reader Word64) (ExprBuilder Id)
```

Comment:
Btw. why don't you define Arbitrary for `Expr` and have it all under `ExprBuilder`?

Reply:
Because that would be even harder! `ExprBuilder` is needed to ensure unique `Id` and `Expr` pairings, which is what drives hash consing: without that, the `Arbitrary` machinery would have to track this information itself, for shrinks too.